crates/kornia-imgproc/src/filter/ops.rs

```diff
@@ -118,6 +122,535 @@ pub fn box_blur_fast<const C: usize>(
+    let full_len = src.rows() * src.cols() * C;
+    let row_len = src.cols() * C;
+
+    {
+        filter_kernel3_row_calc::<C, _>(
+            src_data,
+            src.cols(),
+            &sobel_x,
+            &sobel_y,
+            &mut dx_data[0..row_len],
+            &mut dy_data[0..row_len],
+            0,
+            &|_, dy| dy.max(1) - 1,
+        );
+    }
+    for r in 1..src.rows() - 1 {
+        let row_offset = r * row_len;
+        filter_kernel3_row_calc::<C, _>(
+            src_data,
+            src.cols(),
+            &sobel_x,
+            &sobel_y,
+            &mut dx_data[row_offset..row_offset + row_len],
+            &mut dy_data[row_offset..row_offset + row_len],
+            r,
+            &|r, dy| r + dy - 1,
+        );
+    }
+    {
+        let r = src.rows() - 1;
+        filter_kernel3_row_calc::<C, _>(
+            src_data,
+            src.cols(),
+            &sobel_x,
+            &sobel_y,
+            &mut dx_data[full_len - row_len..full_len],
+            &mut dy_data[full_len - row_len..full_len],
+            r,
+            &|r, dy| (r + dy).min(src.rows()) - 1,
+        );
+    }
+
+    Ok(())
+}
+
+/// Compute the first order image derivative in both x and y using a Sobel operator.
+/// Implement within [separable_filter]
+///
+/// # NOTICE
+///
+/// This only used for benchmark beacuse
+/// `spatial_gradient` require replicate padding, but `separable_filter` use zero padding.
+///
+/// # Arguments
+///
+/// * `src` - The source image with shape (H, W).
+/// * `dst` - The destination image with shape (H, W, 2).
+pub fn spatial_gradient_float_by_separable_filter<const C: usize>(
+    src: &Image<f32, C>,
+    dx: &mut Image<f32, C>,
+    dy: &mut Image<f32, C>,
+) -> Result<(), ImageError> {
+    if src.size() != dx.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dx.cols(),
+            dx.rows(),
+        ));
+    }
+
+    if src.size() != dy.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dy.cols(),
+            dy.rows(),
+        ));
+    }
+
+    let dx_kernel_y = [0.35355339059, 0.70710678118, 0.35355339059]; // 1/sqrt(8), 2/sqrt(8), 1/sqrt(8)
+    let dx_kernel_x = [-0.35355339059, 0.0, 0.35355339059]; // -1/sqrt(8), 0, 1/sqrt(8)
+
+    let dy_kernel_y = [-0.35355339059, 0.0, 0.35355339059]; // -1/sqrt(8), 0, 1/sqrt(8)
+    let dy_kernel_x = [0.35355339059, 0.70710678118, 0.35355339059]; // 1/sqrt(8), 2/sqrt(8), 1/sqrt(8)
+
+    separable_filter(src, dx, &dx_kernel_x, &dx_kernel_y)?;
+    separable_filter(src, dy, &dy_kernel_x, &dy_kernel_y)?;
+
+    Ok(())
+}
+
+/// Compute the first order image derivative in both x and y using a Sobel operator.
+///
+/// # Arguments
+///
+/// * `src` - The source image with shape (H, W).
+/// * `dst` - The destination image with shape (H, W, 2).
+pub fn spatial_gradient_float_parallel<const C: usize>(
```

Comment:
```suggestion
pub fn spatial_gradient_float<const C: usize>(
```
this is not doing doing any explicit parallelization

Reply:
Which implementations should be kept, parallel row by rayon, serial, or more?